ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1318,9 +1310,15 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
         // Find the index segment corresponding to the token indexStartOffset.
         // Get entries starting from the token Key in this index.
         startTimeInMs = time.milliseconds();
-        StoreFindToken newToken =
-            findEntriesFromSegmentStartOffset(storeToken.getOffset(), storeToken.getStoreKey(), messageEntries,
-                new FindEntriesCondition(maxTotalSizeOfEntries), indexSegments, storeToken.getInclusive());
+        StoreFindToken newToken;
+        //if index based token point to the last entry of last index segment, and journal size cleaned up after last log
+        //segment auto close, we should return the original index based token instead of getting from journal.
+        if (onlyContainsHeaderInLastLogSegment(storeToken.getOffset(), indexSegments)) {
```

Comment:
I have doubts on this logic. 

Assume we have such set up

[Index1] [Index2].   [Index3] [Index4].    [Empty Journal]
 [Log Segment 1]      [Log Segment 2]    [new Log segment with only header]

Assuming that last time findEntriesSince returns a IndexBased storeFindToken and the offset is index4.startOffset, but the storeKey is some key in the middle of index4, which means we haven't finished all the IndexEntries in index4. In this case, we shouldn't just assign the storeToken to newToken and move on, we should keep reading the index segments.

Reply:
Updated the logic to check if the storeKey in indexBasedToken is point to the last entry of last index segment or not.